### PR TITLE
Homework#3: Routing

### PR DESCRIPTION
**Pull request by evgeny-sav**

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,9 +1,10 @@
 {
   "presets": [
+    "stage-1",
     "react",
     ["env", {
       "modules": false
     }]
   ],
-  "plugins": ["transform-class-properties"]
+  "plugins": ["transform-class-properties", "transform-async-to-generator"]
```
Comment (mikhail-hatsilau) [inline .babelrc]: I don't think you need `transform-async-to-generator` plugin. ES2017 includes async and await feature. Just specify 2 last  versions of browsers for `env` plugin

File: src/api.js [hunk 2]
```diff
@@ -4,4 +4,10 @@ export default class API {
   static getMovies() {
     return Axios.get('http://localhost:3000/api/v1/movies');
   }
+
+  static getData(searchFor, searchBy) {
+    return Axios
+      .get('http://localhost:3000/api/v1/movies')
+      .then(movies => movies.data.filter(movie => movie[searchBy] === searchFor));
```
Comment (mikhail-hatsilau) [inline src/api.js]: api should only make requests and return responses. Nothing more. All logis should be handled by component

File: webpack.common.js [hunk 3]
```diff
@@ -1,10 +1,13 @@
 const path = require('path');
 const HtmlWebpackPlugin = require('html-webpack-plugin');
+require('babel-core/register');
```
Comment (mikhail-hatsilau) [inline webpack.common.js]: You don't need to include these modules. Moreover you don't need babel-register at all

File: src/index.jsx [hunk 4]
```diff
@@ -1,6 +1,40 @@
...
+                <Route path="/" component={SearchBar} exact />
+                <Route path="/search" component={SearchBar} />
+                <Route path="/film/:id" component={MovieDetails} />
+              </Switch>
+            </Header>
+            <InfoBar>
+              <Route path="/" component={SortBy} />
+            </InfoBar>
+            <Main>
+              <MovieList />
```
Comment (mikhail-hatsilau) [inline src/index.jsx]: You can use MovieList directly in the Main component.

File: src/index.jsx [hunk 5]
```diff
@@ -1,6 +1,40 @@
...
+import InfoBar from './components/InfoBar/InfoBar';
+import SortBy from './components/SortBy/SortBy';
+import Main from './components/Main/Main';
+import NotFound from './components/NotFound/NotFound';
 import './index.scss';
+import MovieList from './components/MovieList/MovieList';
 
-ReactDOM.render(<App />, document.getElementById('root'));
+ReactDOM.render(
+  <Router>
```
Comment (mikhail-hatsilau) [inline src/index.jsx]: I would prefer to move all this jsx to the separate function

File: src/components/SearchBy/SearchBy.jsx [hunk 6]
```diff
@@ -5,29 +5,29 @@ import styles from './SearchBy.scss';
...
+    onHandleSearchBy(val);
   };
 
   return (
     <div className={styles.searchBy}>
       <p className={styles.title}>Search by</p>
       <ul className={styles.list}>
         <li className={cx('item', { active: activeFilter === 'director' })}>
-          <a href="" onClick={e => handelSearch(e, 'director')}>Director</a>
+          <a href="" onClick={e => handelSearchBy(e, 'director')}>Director</a>
```
Comment (mikhail-hatsilau) [inline src/components/SearchBy/SearchBy.jsx:19]: handLe instead of handel :)

File: src/components/SearchBar/SearchBar.jsx [hunk 7]
```diff
@@ -7,27 +9,34 @@ class SearchBar extends Component {
...
-        <form>
+        <div>
           <div className={styles.formTitle}>Find your movie</div>
-          <input type="email" id={styles.searchBar} placeholder="Search" />
-          <SearchBy onHandleSearch={val => this.handelSearch(val)} activeFilter={this.state.searchBy} />
-          <button className={styles.submitBtn} type="submit">
-            Search
-          </button>
-        </form>
+          <input type="text" id={styles.searchBar} placeholder="Search" value={this.state.searchFor} onChange={() => this.handleInputChange} />
```
Comment (mikhail-hatsilau) [inline src/components/SearchBar/SearchBar.jsx]: Use arrow functions as class properties instead

File: src/components/SearchBar/SearchBar.jsx [hunk 8]
```diff
@@ -7,27 +9,34 @@ class SearchBar extends Component {
...
     };
   }
 
-  handelSearch(val = 'title') {
+  handelSearchBy(val = 'title') {
     this.setState({ searchBy: val });
   }
 
+  handleInputChange() {
+    this.setState({ searchFor: 'Batman' }); // e.target.value
```
Comment (mikhail-hatsilau) [inline src/components/SearchBar/SearchBar.jsx]: Why don't you put e.target.value here?

File: src/components/SearchBar/SearchBar.jsx (hunk 7 above)
Comment (mikhail-hatsilau) [inline src/components/SearchBar/SearchBar.jsx]: Destructuring

File: src/components/SearchBar/SearchBar.jsx [hunk 9]
```diff
@@ -7,27 +9,34 @@ class SearchBar extends Component {
     super(props);
     this.state = {
       searchBy: 'title',
+      searchFor: 'Batman',
     };
   }
 
-  handelSearch(val = 'title') {
+  handelSearchBy(val = 'title') {
```
Comment (mikhail-hatsilau) [inline src/components/SearchBar/SearchBar.jsx]: hendLe

File: src/components/NotFound/NotFound.jsx [hunk 10]
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
 
-const NotFound = () => (<h1>NotFound</h1>);
+const NotFound = () => (<h1>404 NotFound</h1>);
```
Comment (mikhail-hatsilau) [inline src/components/NotFound/NotFound.jsx]: You can omit braces. When jsx is one line it doesn't need to be surrounded by braces

File: src/components/MovieList/MovieList.scss [hunk 11]
```diff
@@ -21,7 +21,7 @@
   }
 }
 
-.clearfix {
+.clearfix:after {
```
Comment (mikhail-hatsilau) [inline src/components/MovieList/MovieList.scss]: Let's use `::` for pseudo elements

File: src/components/MovieList/MovieList.jsx [hunk 12]
```diff
@@ -15,12 +23,25 @@ class MovieList extends Component {
...
+
+      MovieList.getData().then((res) => {
+        console.log('update ...');
+        this.setState({
+          movies: [...res],
+        });
       });
-    });
+    }
+    return true;
```
Comment (mikhail-hatsilau) [inline src/components/MovieList/MovieList.jsx]: There is not need to return anything form `componentWillReceiveProps`

File: src/components/MovieList/MovieList.jsx [hunk 13]
```diff
@@ -15,12 +23,25 @@ class MovieList extends Component {
...
-      this.setState({
-        movies: [...movies.data],
+  componentWillReceiveProps({ location }) {
+    if (/search/.test(location.pathname)) {
+      const nextSearchParams = new URLSearchParams(location.search);
+      const prevSearchParams = new URLSearchParams(this.props.location.search);
+      if (location.pathname === this.props.location.pathname &&
+        nextSearchParams.get('searchBy') === prevSearchParams.get('searchBy') &&
+        nextSearchParams.get('searchFor') === prevSearchParams.get('searchFor')) {
+        console.log('no update');
```
Comment (mikhail-hatsilau) [inline src/components/MovieList/MovieList.jsx]: Remove console.log. And I would prefer to compare search params as strings and them convert them to object

File: src/components/MovieList/MovieList.jsx [hunk 14]
```diff
@@ -15,12 +23,25 @@ class MovieList extends Component {
...
+        return false;
```
Comment (mikhail-hatsilau) [inline src/components/MovieList/MovieList.jsx]: Just return

File: src/components/MovieList/MovieList.jsx [hunk 15]
```diff
@@ -15,12 +23,25 @@ class MovieList extends Component {
...
+      }
+
+      MovieList.getData().then((res) => {
```
Comment (mikhail-hatsilau) [inline src/components/MovieList/MovieList.jsx]: async can be used here and let's move it to the separate method

File: src/components/MovieItem/MovieItem.jsx [hunk 16]
```diff
@@ -20,6 +23,7 @@ const MovieItem = ({ title, imageUrl }) => (
 );
 
 MovieItem.propTypes = {
+  movieObj: PropTypes.object.isRequired, // eslint-disable-line
```
Comment (mikhail-hatsilau) [inline src/components/MovieItem/MovieItem.jsx]: You should specify the shape of object. Take a look at `shape` https://facebook.github.io/react/docs/typechecking-with-proptypes.html

File: src/components/MovieDetails/MovieDetails.jsx [hunk 17]
```diff
@@ -1,21 +1,25 @@
 import React from 'react';
+import { Link } from 'react-router-dom';
 import styles from './MovieDetails.scss';
 
 const MovieDetails = () => (
   <div className={styles.movieDetails}>
+    <div className={styles.backBtn}>
+      <Link to="/search" className={styles.btn}>Back to Search</Link>
```
Comment (mikhail-hatsilau) [inline src/components/MovieDetails/MovieDetails.jsx]: You can simply navigate to '/'

File: .babelrc [hunk 18]
```diff
@@ -1,9 +1,10 @@
 {
   "presets": [
+    "stage-1",
```
Comment (mikhail-hatsilau) [inline .babelrc]: You don't need stage-1 preset

File: src/index.jsx [hunk 19]
```diff
@@ -1,6 +1,40 @@
...
+    <App>
+      <Switch>
+        <Route path="/">
+          <div>
+            <Header>
+              <Switch>
+                <Route path="/" component={SearchBar} exact />
+                <Route path="/search" component={SearchBar} />
```
Comment (mikhail-hatsilau) [inline src/index.jsx]: You can remove `/search` route

File: src/components/MovieList/MovieList.jsx (hunk 13 above)
Comment (evgeny-sav) [reply src/components/MovieList/MovieList.jsx]: need some clarifications

File: src/components/MovieList/MovieList.jsx (hunk 15 above)
Comment (evgeny-sav) [reply src/components/MovieList/MovieList.jsx]: need some clarifications

File: src/components/SearchBar/SearchBar.jsx (hunk 7 above)
Comment (evgeny-sav) [reply src/components/SearchBar/SearchBar.jsx]: need some clarifications

File: src/index.jsx (hunk 19 above)
Comment (evgeny-sav) [reply src/index.jsx]: No, I can't cuz when I press Search button I end up on route `/search?...` and on this route `/search` I have to have `search bar`

File: package.json [hunk 20]
```diff
@@ -57,11 +59,13 @@
...
   },
   "devDependencies": {
-    "babel-core": "^6.26.0",
+    "babel-eslint": "^8.0.0",
     "babel-loader": "^7.1.2",
+    "babel-plugin-transform-async-to-generator": "^6.24.1",
     "babel-plugin-transform-class-properties": "^6.24.1",
     "babel-preset-env": "^1.6.0",
     "babel-preset-react": "^6.24.1",
+    "babel-preset-stage-1": "^6.24.1",
```
Comment (mikhail-hatsilau) [inline package.json:68]: remove stage-1 preset

File: src/components/MovieItem/MovieItem.jsx [hunk 21]
```diff
@@ -1,15 +1,18 @@
 import React from 'react';
+import { Link } from 'react-router-dom';
 import classNames from 'classnames/bind';
 import PropTypes from 'prop-types';
 import styles from './MovieItem.scss';
 
 const cx = classNames.bind(styles);
 
-const MovieItem = ({ title, imageUrl }) => (
+const MovieItem = ({ title, imageUrl, movieObj }) => (
```
Comment (mikhail-hatsilau) [inline src/components/MovieItem/MovieItem.jsx:9]: If you pass the whole movie than you don't need to keep title and imageUrl props separately. You can take them from movieObj

File: src/api.js [hunk 22]
```diff
@@ -4,4 +4,8 @@ export default class API {
   static getMovies() {
     return Axios.get('http://localhost:3000/api/v1/movies');
   }
+
+  static getData() {
```
Comment (mikhail-hatsilau) [inline src/api.js:8]: Let's rename it to smth like searchFilms. And it should get to params which you pass: searchFor and searchBy

File: src/components/SearchBar/SearchBar.jsx [hunk 23]
```diff
@@ -7,27 +9,35 @@ class SearchBar extends Component {
...
-        <form>
+        <div>
           <div className={styles.formTitle}>Find your movie</div>
-          <input type="email" id={styles.searchBar} placeholder="Search" />
-          <SearchBy onHandleSearch={val => this.handelSearch(val)} activeFilter={this.state.searchBy} />
-          <button className={styles.submitBtn} type="submit">
-            Search
-          </button>
-        </form>
+          <input type="text" id={styles.searchBar} placeholder="Search" value={searchFor} onChange={e => this.handleInputChange(e)} />
```
Comment (mikhail-hatsilau) [inline src/components/SearchBar/SearchBar.jsx:30]: if you have used arrow function for handleInputChange then you can directly pass it to the onChange callback: `onChange={ this.handleInputChange }`

File: src/components/SearchBar/SearchBar.jsx [hunk 24]
```diff
@@ -7,27 +9,35 @@ class SearchBar extends Component {
...
+          <SearchBy onHandleSearchBy={val => this.handleSearchBy(val)} activeFilter={searchBy} />
```
Comment (mikhail-hatsilau) [inline src/components/SearchBar/SearchBar.jsx:31]: the same is here: `onHandleSearchBy={ this,handleSearchBy }`

**Comment by mikhail-hatsilau** [comment]
All comment should be fixed :)

File: src/api.js (hunk 22 above)
Comment (evgeny-sav) [reply src/api.js:8]: need some explanations

**Pull request merged**